### PR TITLE
[NEW] Subgroups

### PR DESCRIPTION
**Pull request by johnmberger**
@RocketChat/core 

Introduces **subgroups**, which are child rooms that can be created from channels or private groups. These subgroups are ideal for conversations that are tangentially related to a parent room. Subgroups can be created via a menu on the flextab bar.

_creating a subgroup_
screen shot 2017-11-15 at 11 32 49 am

Users have the option of collapsing a parent room's subgroups to reduce clutter on the sidebar.

_uncollapsed_
screen shot 2017-11-15 at 11 26 00 am

_collapsed_
screen shot 2017-11-15 at 11 31 32 am

Users also have the option to have subgroups collapsed by default when they log in.
screen shot 2017-11-15 at 11 30 52 am

### CODE REVIEW

**Comment by gdelavald** [comment]
This is pretty awesome, thanks for submitting the PR. We'll evaluate and provide feedback as soon as we can.

**Comment by HammyHavoc** [comment]
This looks awesome.

**Comment by lunitic** [comment]
@johnmberger Nice work !!

@core team any idea in which release this will be added ?s

**Comment by Sing-Li** [comment]
@johnmberger  awesome.  And thanks!

**Comment by mrsimpson** [comment]
@johnmberger truly awesome! 
It's some kind of threading which has been asked for for a very very very long time and very very often (see #1112, #6888). We implemented something similar in our fork as well an I'd love to see this merged.
However, I'm afraid it will not be that likely to get this merged: Threading will be a core and very important feature to Rocket.Chat and might even bring some uniqueness to it. Thus, it's such a long lasting decision which even @engelgabriel seems not to be able to make.

### A pledge to @RocketChatApp 
In order to make it possible for developers implementing solutions on their own -  LIKE @johnmberger and myself, it would be very helpful to at least know your cornerstones: How will the data model look like and are there any defined interfaces already? This would make it possible for us to develop our custom threading workarounds and switch to the "real" threading once it's available.

File: packages/rocketchat-lib/server/methods/addUsersToRoom.js [hunk 1]
```diff
@@ -58,8 +58,34 @@ Meteor.methods({
...
 			RocketChat.addUserToRoom(data.rid, newUser, user);
+
+			if (room.subGroup) {
+				let sub = RocketChat.models.Subscriptions.update({
+					'rid': data.rid
+				},
+				{
+					$set:
+						{
+							'subGroup': true,
```
Comment (mrsimpson) [inline packages/rocketchat-lib/server/methods/addUsersToRoom.js]: It was e. g. interesting to know whether the thread will be a separate room type or remain in the same type as the parent

File: packages/rocketchat-lib/server/methods/addUsersToRoom.js [hunk 2]
```diff
@@ -58,8 +58,34 @@ Meteor.methods({
...
+							'subGroupName': room.subGroupName,
+              'originalRoomId': room.originalRoomId
```
Comment (mrsimpson) [inline packages/rocketchat-lib/server/methods/addUsersToRoom.js]: which property shall be used to form the hierarchy - or whether a hierarchy shall exist after all and not only some sort of "tagging"

Comment (mrsimpson) [reply packages/rocketchat-lib/server/methods/addUsersToRoom.js]: Also, it was interesting to know whether the subroom is just a room within a room or whether it is started from a message (which actually made it a thread)

**Comment by kenton-r** [comment]
It would be great if these conflicts could be resolved. Having sub channels would be a great benefit. If it is not much work to resolve the issues. it looks like there has already been some work done to get this feature working.

**Comment by csimpi** [comment]
Is this a dead thread? This function would be great.

**Comment by johnmberger** [comment]
@csimpi - think so unfortunately. Rocket's been going back and forth about message threading for awhile and probably won't merge this in until they make up their mind.

**Comment by csimpi** [comment]
Aww, it's a same. Nowadays the biggest issue with the productivity team chats is the searching in the history all the time. I don't understand why nobody is willing to implement a solution like this, it would be so much easier If the users would be able to tag their messages if that's necessary. 
I would use this for my personal life, for example my wife used to send me list for the grocery, instagram names to follow, movies and tv show to watch, then we can't find anything because the lot of data in the same thread. I don't wanna open channels for every small topic. This would be a very necessary function but some reason Rocketchat's developers are getting frustrated by this feature request, I don't know why, they think this is against the chat, but that's not true, this would be the next generation of chatting/discussing. I'm thinking about a startup right now. How's in? :))

**Comment by csimpi** [comment]
Yes, I completely agree, there should be an API for plugins, it would be awesome to solve this in the house :)

**Comment by yevgenydevine** [comment]
@Sing-Li @brunosquadros this is the top most important feature. It is vital for platform organization and community management.
Implement it at least on the most basic level, make it optional, but let people have it.

We, and other people we know, have gone through severe complications to have a half-decent, semi-working workaround for this feature. If it were integrated on the most basic level, it wouldn't be a problem to adjust it however one wants.

**Comment by peterkogit** [comment]
As I understand, the function does not work?

**Comment by engelgabriel** [comment]
> We have decided that was better to delay our v1 release and get the Threads done properly than get the feature out the way it was initially implemented on the develop branch. We have "rebranded" our current implementation of Threads to Discussions (was it creates a new room type) and will implement Threads exactly the same way as Slack does.
>
> In our vision there are 2 different use cases on how to break out of the main conversation: one very easy and organic, with a single click for small side talk, and other, much more structured, that creates a new conversation with its own life cycle.
>
> We have tried to have a single solution for both, and as a result, no one was happy with it.
So we have given up on having a single approach for both use cases and will implement both: the Slack-like industry standard and our brand new one, called Discussions.

https://github.com/RocketChat/Rocket.Chat/issues/13843

**Comment by engelgabriel** [comment]
Thank you @johnmberger for the hard work on this PR and @mrsimpson for the feedback and deep technical considerations about the underlying data model. We have used this code as the bases for our implementation. 

The result is the initial implementation added via PR #13541 by @ggazzo

We will still use this PR as a reference in the future as we plan to improve the navigation by moving discussion under their parent channels. We will wait for more community input before we move this forward.

Thank you again for helping us build such a complex and highly demanded feature.

**Comment by 247Kittichai** [comment]
I would like to know about Subgroups.
rocket.chat Version 2.0.0  Can I create subgroups ?
Thank you.

**Pull request closed without merging**